### ISSUE TITLE
Emmet: "Wrap with abbreviation support"

### ISSUE THREAD

**Issue by Vinnl** [zed-industries/zed#15588]
### Check for existing issues

- [X] Completed

### Describe the feature

One very useful Emmet feature is [wrap with abbreviation](https://docs.emmet.io/actions/wrap-with-abbreviation/): enter an Emmet abbreviation, and then it gets expanded to include the current selection.

### If applicable, add mockups / screenshots to help present your vision of the feature

See demo at https://docs.emmet.io/actions/wrap-with-abbreviation/.

**Comment by tuzemec** [comment]
This is probably the feature that I miss the most from Sublime

**Comment by codegod100** [comment]
I have the emmet extension, is this functionality not included in that?

**Comment by getflourish** [comment]
Came here to say I miss this! Any extension references to look at that do something similar?

**Comment by getflourish** [comment]
Interesting!

**Comment by arscynic** [comment]
> Interesting!

My code only works for wrapping tags around one continuous selection though. Not for multiple selections nor actual Emmet abbreviations. I tried to implement the latter with `keyboard.send_keys("%s<tab>" % text_input[1] + "%s" % text_selected)`, but no success yet.

**Comment by tuzemec** [comment]
As far as I understand "wrap in abbreviation" is not a LSP command and the LSP integration needs to take care of this functionality.
This is how it is implemented in nvim: https://github.com/olrtg/nvim-emmet/blob/main/lua/nvim-emmet/init.lua

I may be completely wrong here, but for this to work, Zed would need to provide some kind of input field or interface where the abbreviation can be defined. I'm not familiar with the internals, but this doesn't seem like a simple task.

**Comment by fmarkulin** [comment]
Highly miss this from VS Code in web development

**Comment by riesemann** [comment]
As a web developer I miss that functionality too.
... zed is awesome!

**Comment by ddbrierton** [comment]
Is this issue related to #10394? Or is it just coincidence? It would be good to know if there is an underlying issue blocking progress on both of these tickets that we could track.

I also wonder if the fact that so much of the snippet spec is not implemented is also related? See #13301.

If Emmet's "wrap with abbreviation" is not supported, AND Vim Surround's surround with tag is not supported, AND the `TM_SELECTED_TEXT` snippet variable is not supported, then it would appear that there is no straightforward way to wrap some text in an HTML tag in Zed, which strikes me as fairly basic editor functionality in this day and age.

To be honest, I'm surprised that this has not been made a higher priority.

**Comment by tuzemec** [comment]
Yeah, https://github.com/zed-industries/zed/issues/10394 seems very related, but it looks like the implementation was abandoned. :(

And this one doesn’t have many votes, so it’s not surfacing among the top issues.

**Comment by 4c656f** [comment]
Hi everyone,

Just wanted to share a partial workaround I've put together for wrapping selected text with an HTML tag using Zed's keybinding system.

It doesn't implement full Emmet abbreviation support, but it allows you to wrap the currently selected text with a simple tag by manually typing the tag name.

Here's the keybinding you can add to your `keymap.json` (assuming you're using `@k` for this):

```json
{
    "context": "Editor && vim_mode == visual && !VimWaiting && !VimObject",
    "bindings": {
      "@ k": ["workspace::SendKeystrokes", "d i <>\n</> escape k p h i"]
    }
  }
```

**How to use:**

1.  Select the text in a visual mode you want to wrap.
2.  Press your bound key combination (e.g., `@k`).
3.  Type the name of the tag you want to use (e.g., `div`, `span`, `p`). The cursor will be placed inside the opening angle brackets `<>`.

**Limitations:**

*   This is not Emmet support.
*   It only handles a single, simple tag at a time.
*   It requires manually typing the tag name for both the opening and closing tags (partially automated by placing cursor for opening, but closing is manual).
*   (Note: This uses Vim mode keystrokes via `SendKeystrokes`, assuming that's the intended context or how `SendKeystrokes` works for the user).

Hope this might be helpful for some while waiting for a more complete solution!

**Comment by ddbrierton** [comment]
I'm a little concerned that this issue doesn't have the "languages" label. If you look at Zed's road map

https://zed.dev/roadmap

there are "language-specific issues" that are deal-breakers for a 1.0 release. As "Web" is listed there, I would've thought that being able to add tags to things was pretty fundamental. There are so many issues that it is easy for things to fall between the cracks, especially when the focus is elsewhere such as on agents and the debugger.

I don't know how one should try to bring this to the attention of the Zed team.

**Comment by quinncomendant** [comment]
FWIW, Claude Sonnet 4 understands Emmet abbreviations, so you can do this with the [inline assistant](https://zed.dev/docs/ai/inline-assistant). In fact, you can mix Emmet shorthand with natural language. [Here’s a screen recording](https://send.strangecode.com/f/screen-recording-2025-06-14-at-11-30-43.mp4) to demo this:

1. Start with these lines selected and invoke inline assistant with `ctrl-return`:
```
one
two
three
```
2. Give inline assistant the instructions: `ul>li.item first one is .selected` and hit return, then return again to accept.
3. The result is:
```html
<ul>
    <li class="item selected">one</li>
    <li class="item">two</li>
    <li class="item">three</li>
</ul>
```

**Comment by yepes** [comment]
Native support would be desirable and better. I don't see why we should rely on AI to stuff that could be done for years in other programs.

**Comment by olrtg** [comment]
Hello guys, author of the [emmet-language-server](https://github.com/olrtg/emmet-language-server) here. The language server has an off-spec endpoint to implement this feature (source [here](https://github.com/olrtg/emmet-language-server/blob/279be108725fb391c167690b697ce154fd32657b/index.ts#L178-L197)), since is off-spec (due to code actions not supporting user input) we have to make a custom call to the language server like in [nvim-emmet](https://github.com/olrtg/nvim-emmet/blob/cde4fb2968704aae5c18b7f8a9bc2508767bb78d/lua/nvim-emmet/init.lua#L27-L39), but in Zed we don't have an API to talk to LSPs through extensions yet. So until this is available to us users, I don't think this will get implemented.

Another alternative maybe is to have some sort of way through commands (like `lsp:SendRequest`) and keybindings to make this custom call. But that's likely a bad idea, or weird, idk.

**Comment by henrikklee** [comment]
My most missed feature in Zed. Support for this would be great!

**Comment by AwzemCoffee** [comment]
I'll further ask for this! Zed is a _wonderful_ editor and I'd absolutely love to see support for this added soon.

**Comment by paulmsmith** [comment]
How likely is this to ever be implemented?

**Comment by rcaredu** [comment]
How disappointing that this has not been implemented. Deal-breaker.

**Comment by Nemesis19** [comment]
+1 for this feature, essential in the daily workflow.

**Comment by chances** [comment]
> How likely is this to ever be implemented?

https://github.com/zed-industries/zed/issues/15588#issuecomment-4411619453

At least this has climbed to [page 2](https://github.com/zed-industries/zed/issues?q=is%3Aissue%20state%3Aopen%20sort%3Areactions-%2B1-desc&page=2) of the most up-voted issues, but it _is_ the **most** upvoted of all the [Emmet issues](https://github.com/zed-industries/zed/issues?q=state%3Aopen%20label%3A%22area%3Atooling%2Femmet%22%20sort%3Areactions-%2B1-desc).

---

What do we have to do to get this on the front-burner? Is it a question of resources?

**Comment by SomeoneToIgnore** [comment]
Hey, I have worked on in in https://github.com/zed-industries/zed/pull/63383

So far things seem to be working, but I have never personally used Emmet hence might have missed some workflow aspects or other things: and for that I would appreciate any feedback and notes, here or in the PR.

The prebuilt binaries are available at the bottom of https://github.com/zed-industries/zed/actions/runs/33204650935?pr=63383 page, and the branch itself can be built locally, if needed.
